runtime/cudaq/platform/default/rest_server/RemoteRESTQPU.cpp

```diff
@@ -0,0 +1,296 @@
+      moduleOp.push_back(func.clone());
+
+      if (args) {
+        cudaq::info("Run Quake Synth.\n");
+        PassManager pm(&m_mlirContext);
+        pm.addPass(cudaq::opt::createQuakeSynthesizer(name, args));
+        if (failed(pm.run(moduleOp)))
+          throw std::runtime_error("Could not successfully apply quake-synth.");
+      }
+
+      // Client-side passes
+      if (!clientPasses.empty()) {
+        PassManager pm(&m_mlirContext);
+        std::string errMsg;
+        llvm::raw_string_ostream os(errMsg);
+        const std::string pipeline =
+            std::accumulate(clientPasses.begin(), clientPasses.end(),
+                            std::string(), [](const auto &ss, const auto &s) {
+                              return ss.empty() ? s : ss + "," + s;
+                            });
+        if (failed(parsePassPipeline(pipeline, pm, os)))
+          throw std::runtime_error(
+              "Remote rest platform failed to add passes to pipeline (" +
+              errMsg + ").");
+
+        if (failed(pm.run(moduleOp)))
+          throw std::runtime_error(
+              "Remote rest platform: applying IR passes failed.");
+      }
+      std::string mlirCode;
+      llvm::raw_string_ostream outStr(mlirCode);
+      mlir::OpPrintingFlags opf;
+      opf.enableDebugInfo(/*enable=*/true,
+                          /*pretty=*/false);
+      moduleOp.print(outStr, opf);
+      request.setCode(mlirCode);
+    }
+
+    request.entryPoint = name;
+    request.simulator = m_simName;
+    request.seed = cudaq::get_random_seed();
+    // Don't let curl adding "Expect: 100-continue" header, which is not
+    // suitable for large requests, e.g., bitcode in the JSON request.
+    //  Ref: https://gms.tf/when-curl-sends-100-continue.html
+    std::map<std::string, std::string> headers{
+        {"Expect:", ""}, {"Content-type", "application/json"}};
+    json requestJson = request;
+    auto resultJs = m_client.post(m_url, "job", requestJson, headers);
+    resultJs.get_to(*executionContext);
+  }
+
+  void setExecutionContext(cudaq::ExecutionContext *context) override {
+    cudaq::info("RemoteSimulatorQPU::setExecutionContext QPU {}", qpu_id);
+    std::scoped_lock<std::mutex> lock(m_contextMutex);
+    m_contexts.emplace(std::this_thread::get_id(), context);
+  }
+
+  void resetExecutionContext() override {
+    cudaq::info("RemoteSimulatorQPU::resetExecutionContext QPU {}", qpu_id);
+    std::scoped_lock<std::mutex> lock(m_contextMutex);
+    m_contexts.erase(std::this_thread::get_id());
+  }
+};
+
+class RemoteSimulatorQuantumPlatform : public cudaq::quantum_platform {
+  std::unique_ptr<MLIRContext> m_mlirContext;
+  std::unordered_map<std::thread::id, std::size_t> m_threadIdToQpuId;
+
+public:
+  ~RemoteSimulatorQuantumPlatform() = default;
+  RemoteSimulatorQuantumPlatform() : m_mlirContext(cudaq::initializeMLIR()) {
+    platformNumQPUs = 0;
+  }
+  virtual void kernelEnter(const std::string &kernelName,
```

Comment:
Why didn't just overriding launchKernel not work?

Reply:
The new platform impl overrides `launchKernel` for MLIR mode where kernel entry point is redirected to `altLaunchKernel`. 
One of the requirements is to have support for library mode in this remote execution platform as well. The approach is to send on the library mode LLVM bitcode then JIT and execute the kernel on the server. Hence, I'm adding this function to notify the platform that the library-mode kernel is about to be launch, need to call `launchKernel`.